packages/tools/piw-utils-internal/src/utils/PopupUtils.ts

```diff
@@ -0,0 +1,264 @@
+    if (pointContainer) {
+        do {
+            if (pointContainer === element) {
+                return true;
+            } else {
+                pointContainer = pointContainer.parentElement as HTMLElement;
+            }
+        } while (pointContainer.parentElement);
+    }
+    return false;
+}
+
+export function isElementPartiallyOffScreen(dynamicWindow: Window, rect: DOMRect): boolean {
+    return (
+        rect.x < 0 ||
+        rect.y < 0 ||
+        rect.x + rect.width > dynamicWindow.document.documentElement.clientWidth ||
+        rect.y + rect.height > dynamicWindow.document.documentElement.clientHeight
+    );
+}
+
+export function moveAbsoluteElementOnScreen(
+    dynamicWindow: Window,
+    element: HTMLElement,
+    boundingRect: DOMRect
+): DOMRect {
+    if (boundingRect.x < 0) {
+        const leftValue = Math.round(getPixelValueAsNumber(element, "left") - boundingRect.x);
+        element.style.left = leftValue + "px";
+        boundingRect.x += leftValue;
+    }
+    if (boundingRect.y < 0) {
+        const topValue = Math.round(getPixelValueAsNumber(element, "top") - boundingRect.y);
+        element.style.top = topValue + "px";
+        boundingRect.y += topValue;
+    }
+    if (boundingRect.x + boundingRect.width > dynamicWindow.document.documentElement.clientWidth) {
+        const rightValue = Math.round(
+            getPixelValueAsNumber(element, "right") +
+                (boundingRect.x + boundingRect.width - dynamicWindow.document.documentElement.clientWidth)
+        );
+        element.style.right = rightValue + "px";
+        boundingRect.x -= rightValue;
+    }
+    if (boundingRect.y + boundingRect.height > dynamicWindow.document.documentElement.clientHeight) {
+        const bottomValue = Math.round(
+            getPixelValueAsNumber(element, "bottom") +
+                (boundingRect.y + boundingRect.height - dynamicWindow.document.documentElement.clientHeight)
+        );
+        element.style.top = "unset"; // Unset top defined in PopupMenu.scss
+        element.style.bottom = bottomValue + "px";
+        boundingRect.y -= bottomValue;
+    }
+    return boundingRect;
+}
+
+export function handleOnClickOutsideElement(ref: RefObject<HTMLDivElement>, handler: () => void): void {
+    useEffect(() => {
+        const listener = (event: MouseEvent & { target: Node | null }): void => {
+            if (!ref.current || ref.current.contains(event.target)) {
+                return;
+            }
+            handler();
+        };
+        ref.current?.ownerDocument.addEventListener("mousedown", listener);
+        ref.current?.ownerDocument.addEventListener("touchstart", listener);
+        return () => {
+            ref.current?.ownerDocument.removeEventListener("mousedown", listener);
+            ref.current?.ownerDocument.removeEventListener("touchstart", listener);
+        };
+    }, [ref, handler]);
+}
```

Comment:
I see that this file is a copy of popup menu widget utils. But in popup menu we don't use them, it still uses local copy of the utils. I would expect this to be gone from the popup menu widget and be imported from here.

Ideally we need to re-use UI components for popup menus, not copy-paste those from widget to widget.

Reply:
Not for now